src/RobustLeastSquares.jl

```diff
@@ -1,5 +1,307 @@
+# (C3 continuous but not analytically smooth)
+"""The (convex) "fair" estimator switches from between quadratic and linear cost/loss function at a certain cutoff, and is C3 but non-analytic."""
+immutable FairEstimator <: MEstimatorConvex; width::Float64; end
+m_estimator_rho(r,est::FairEstimator) = est.width^2 * (abs(r)/est.width - log(1 + abs(r)/est.width))
+m_estimator_psi(r,est::FairEstimator) = r ./ (1.0 + abs(r)/est.width)
+m_estimator_weight(r,est::FairEstimator) = 1.0 ./ (1.0 + abs(r)/est.width)
+m_estimator_sqrtweight(r,est::FairEstimator) = 1.0 ./ sqrt(1.0 + abs(r)/est.width)
+
+# The Cauchy (or Lorentzian) estimator. Non-convex with logarithmically-diverging
+# cost function means that wild outliers are mostly ignored, while still having a
+# somewhat-robust field-of-influence
+"""The non-convex Cauchy estimator switches from between quadratic behaviour to logarithmic tails. This rejects outliers but may result in mutliple minima."""
+immutable CauchyEstimator <: MEstimatorNonConvex; width::Float64; end
+m_estimator_rho(r,est::CauchyEstimator) = 0.5*est.width^2 * log(1 + r.*r/(est.width*est.width))
+m_estimator_psi(r,est::CauchyEstimator) = r ./ (1.0 + r.*r/(est.width*est.width))
+m_estimator_weight(r,est::CauchyEstimator) = 1.0 ./ (1.0 + r.*r/(est.width*est.width))
+m_estimator_sqrtweight(r,est::CauchyEstimator) = 1.0 ./ sqrt(1.0 + r.*r/(est.width*est.width))
+
+
+"""
+A custom type for applying an estimators to only a subset of the range
+(uses L2 for any unspecified range)
+"""
+immutable SubsetEstimatorConvex{T<:MEstimatorConvex} <: MEstimatorConvex
+    rng::UnitRange{Int64}
+    est::T
+end
+immutable SubsetEstimator{T<:MEstimator} <: MEstimatorNonConvex
+    rng::UnitRange{Int64}
+    est::T
+end
+
+function m_estimator_rho{T}(r,est::Union{SubsetEstimatorConvex{T},SubsetEstimator{T}})
+    rho = 0.5*r.^2
+    rho[est.rng] = m_estimator_rho(r[est.rng],est.est)
+    return rho
+end
+function m_estimator_psi{T}(r,est::Union{SubsetEstimatorConvex{T},SubsetEstimator{T}})
+    psi = r
+    psi[est.rng] = m_estimator_psi(r[est.rng],est.est)
+    return psi
+end
+function m_estimator_weight{T}(r,est::Union{SubsetEstimatorConvex{T},SubsetEstimator{T}})
+    w = ones(size(r))
+    w[est.rng] = m_estimator_weight(r[est.rng],est.est)
+    return w
+end
+function m_estimator_sqrtweight{T}(r,est::Union{SubsetEstimatorConvex{T},SubsetEstimator{T}})
+    w = ones(size(r))
+    w[est.rng] = m_estimator_sqrtweight(r[est.rng],est.est)
+    return w
+end
+
+
+"""
+A custom type for applying two estimators to different ranges
+(uses L2 for any unspecified range)
+"""
+immutable DualEstimatorConvex{T1<:MEstimatorConvex,T2<:MEstimatorConvex} <: MEstimatorConvex
+    rng1::UnitRange{Int64}
+    est1::T1
+    rng2::UnitRange{Int64}
+    est2::T2
+end
+immutable DualEstimator{T1<:MEstimator,T2<:MEstimator} <: MEstimatorNonConvex
+    rng1::UnitRange{Int64}
+    est1::T1
+    rng2::UnitRange{Int64}
+    est2::T2
+end
```

Comment:
Two things here.
* Can we get away with a single type here (what's the current function of convex vs nonconvex?)
* I think we need a better name than DualEstimator - Dual has strong mathematical connotations.  `EstimatorPair` ?